llama_index/program/predefined/evaporate.py

```diff
@@ -0,0 +1,111 @@
+        orig_prompt = Prompt(df_parser_template_str)
+        column_schema = ",".join(fields_to_extract)
+        new_prompt = Prompt.from_prompt(
+            orig_prompt.partial_format(
+                column_schema=column_schema,
+            )
+        )
+
+        pydantic_program = pydantic_program_cls.from_defaults(
+            DataFrameRowsOnly, new_prompt.original_template, **program_kwargs
+        )
+        self._validate_program(pydantic_program)
+        self._pydantic_program = pydantic_program
+        self._extractor = EvaporateExtractor()
+        self._field_fns: Dict[str, str] = {}
+        self._fields = fields_to_extract
+        self._training_key = training_key
+        self._infer_key = infer_key
+
+    def _validate_program(self, pydantic_program: BasePydanticProgram) -> None:
+        if pydantic_program.output_cls != DataFrameRowsOnly:
+            raise ValueError(
+                "Output class of pydantic program must be `DataFramRowsOnly`."
+            )
+
+    def _fit(self, nodes: List[BaseNode], field: str) -> str:
+        """Given the input Nodes and fields, synthesize the python code."""
+        fn = self._extractor.extract_fn_from_nodes(nodes, field)
+        logger.debug(f"Extracted function: {fn}")
+        return fn
+
+    def _inference(self, nodes: List[BaseNode], fn_str: str, field_name: str) -> str:
+        """Given the input, call the python code and return the result."""
+        results = self._extractor.run_fn_on_nodes(nodes, fn_str, field_name)
+        logger.debug(f"Results: {results}")
+        return str(results)
+
+    @classmethod
+    def from_defaults(
+        cls,
+        fields_to_extract: List[str],
+        pydantic_program_cls: Optional[Type[BaseLLMFunctionProgram]] = None,
+        df_parser_template_str: str = DEFAULT_ROWS_DF_PARSER_TMPL,
+        **kwargs: Any,
+    ) -> "EvaporateProgram":
+        """Rows DF output parser."""
+        pydantic_program_cls = pydantic_program_cls or OpenAIPydanticProgram
+
+        return cls(
+            pydantic_program_cls,
+            fields_to_extract,
+            df_parser_template_str=df_parser_template_str,
+            **kwargs,
+        )
+
+    @property
+    def output_cls(self) -> Type[DataFrameRowsOnly]:
+        """Output class."""
+        return DataFrameRowsOnly
+
+    def __call__(self, *args: Any, **kwds: Any) -> DataFrameRowsOnly:
+        """Call evaporate on training and inference data. Inputs should be two
+        lists of nodes, using the keys `training_data` and `infer_data`."""
+        for field in self._fields:
+            self._field_fns[field] = self._fit(kwds[self._training_key], field)
+        infer_dict = {}
+        for field in self._fields:
+            infer_dict[field] = self._inference(
+                kwds[self._infer_key], self._field_fns[field], field
+            )
+        infer_results = str(infer_dict)
+        result = self._pydantic_program(**{"input_str": infer_results})
```

Comment:
i actually don't know why we're using a pydantic program here. is the only point to just parse a dictionary into a DataFrame? You don't need an LLM to do this since you already have the structured information in `infer_dict` 

Also I don't know why we're using `DataFrameRowsOnly` vs `DataFrameRow` since it's just a single row. You can construct `DataFrameRow` through code instead of using the LLM